### ISSUE TITLE
Cross-platform design of the tasktray icon launcher menu

### ISSUE THREAD

**Issue by jamalex** [learningequality/installers#6]
The goal is to not require use of the command line on any platform, for performing basic tasks like:
- Starting the KA Lite server
- Stopping the KA Lite server
- Opening the correct URL in the browser for accessing KA Lite
- Basic settings regarding the launching of the server (e.g. whether to launch at boot)

While this will likely need to be implemented in OS-specific ways, we should keep the design (icon, menu options, behavior) consistent across OS implementations. This issue is for discussing that design.

**Comment by jamalex** [comment]
Looking pretty good! It might be nice to have separate "Start server" and "Stop server" entries, and just grey out the one that's currently not needed. "Configuration" might be better as "Settings".

"Run KA Lite when the user *logs in*"

Remove "Minimize to tray" since there's now only a tray. :)

In terms of user experience, it would be good to get input from @jtamiace. I'm wondering whether users (especially ones using it in single-computer loopback mode, rather than for connecting other computers to) will know what we mean by "Start server". Other than using system resources, having the server always running when the task tray app is running almost makes sense. And then, clicking the icon could launch the browser directly, and right-clicking could show a settings-focused menu. (These are all just brainstorming ideas, not direct suggestions. It would be good to get a UX group to brainstorm this.) 

> Should include the "about" option?

The less we can put in here, the better, so my vote would be no.

> OFF TOPIC - This would probably go to a different issue but I was wondering if we should support an option to change the language. I think that would make sense to have the GUI available in different languages - internationalization system.

Fantastic idea -- yes, we definitely should. Could you open an issue in this repo about that?

**Comment by jamalex** [comment]
Another UX question -- if we now only have the task tray icon, and no pop-up GUI, it might be easy for the user not to notice that it's there, the first time. Perhaps we can pop up a bubble when it launches, telling the user that it's there, or something?

**Comment by ruimalheiro** [comment]
@jamalex thanks a lot for the input!

> Another UX question -- if we now only have the task tray icon, and no pop-up GUI, it might be easy for the user not to notice that it's there, the first time. Perhaps we can pop up a bubble when it launches, telling the user that it's there, or something?

This can be done without any problem. You can send a message when you want (and of the type you want) as a notification to the tray menu. So we can achieve that easily, at least in Windows. :+1:

**Comment by ruimalheiro** [comment]
I forgot to ask. Should we still show the version number under the tray menu?

**Comment by jamalex** [comment]
> I forgot to ask. Should we still show the version number under the tray menu?

It would increase the complexity, since the software can (soon) be updated without updating the task tray icon, and so it would need to poll for updates to version.py. I think we can leave that to the web interface, so: not needed here.

**Comment by ruimalheiro** [comment]
@jamalex sounds good to me, but you know that the GUI actually gets the version number from version.py (so it is nothing new here. )

If you want me to remove it, that is fine to me. (Just wanting to be sure that you are aware of the existence of the functionality) :+1:

**Comment by jamalex** [comment]
Yep! In the past, that approach worked fine, as the only way to update KA Lite was with a new installer. Soon, there will be an internal update mechanism, so KA Lite might change version without the GUI restarting.

**Comment by ruimalheiro** [comment]
Oh, now I understand your point! OK, will remove that! :+1:

**Comment by ruimalheiro** [comment]
@jamalex @aronasorman @cpauya 

Update: You can now pass normal char* strings intead of the need for LPWSTR.

A better example:

```
#include "fle_win32_framework.h"

// Declare global stuff that you need to use inside the functions.
fle_TrayWindow * window;

fle_TrayMenuItem * menu1;
fle_TrayMenuItem * menu2;
fle_TrayMenuItem * menu3;
fle_TrayMenuItem * menu4;
fle_TrayMenuItem * menu5;
fle_TrayMenuItem * menu6;
fle_TrayMenuItem * menu7;


void trayMenuAction1()
{
	if(!runShellScript("start.bat", NULL, "ka-lite\\"))
	{
		// Handle error.
		printConsole("Failed to run the script.\n");
	}
	else
	{
		menu1->disable();
		menu2->enable();
		printConsole("The script was run successfully.\n");
	}
}

void trayMenuAction2()
{
	if(!runShellScript("stop.bat", NULL, "ka-lite\\"))
	{
		// Handle error.
		printConsole("Failed to run the script.\n");
	}
	else
	{
		menu2->disable();
		menu1->enable();
		printConsole("The script was run successfully.\n");
	}
}

void trayMenuAction3()
{
	if(menu4->isChecked())
	{
		menu4->uncheck();
		menu5->enable();
	}
	else
	{
		menu4->check();
		menu5->disable();
	}
}

void trayMenuAction4()
{
	printConsole("I am exiting KA Lite.\n");
}

void statusFunction()
{
	// The code here runs on each frame of the window main loop.
	// We can handle things like checking if the server is online and controlling the state of each component.
	// Basically, the automated stuff.
}

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
	window = new fle_TrayWindow(&hInstance);
	window->setTrayIcon("images\\logo48.ico");
	window->setStatusFunction(&statusFunction);

	menu1 = new fle_TrayMenuItem("Start Server.", &trayMenuAction1);
	menu2 = new fle_TrayMenuItem("Stop Server.", &trayMenuAction2);
	menu3 = new fle_TrayMenuItem("Options", NULL);
	menu4 = new fle_TrayMenuItem("Run KA Lite when the user logs in.", &trayMenuAction3);
	menu5 = new fle_TrayMenuItem("Run KA Lite at system startup.", NULL);
	menu6 = new fle_TrayMenuItem("Auto-start server when KA Lite is run.", NULL);
	menu7 = new fle_TrayMenuItem("Exit KA Lite.", &trayMenuAction4);

	menu3->setSubMenu();
	menu3->addSubMenu(menu4);
	menu3->addSubMenu(menu5);
	menu3->addSubMenu(menu6);
	
	window->addMenu(menu1);
	window->addMenu(menu2);
	window->addMenu(menu3);
	window->addMenu(menu7);

	window->show();

	return 0;
}
```

Output:
![image](https://cloud.githubusercontent.com/assets/1102381/5793513/36bb9500-9f3e-11e4-94e8-d284dc143662.png)

**Comment by jamalex** [comment]
Would be good to get the opinions of some of our UX folks (e.g. @howdyitshelena @kkmach @crystalkwok @jtamiace) about the wording -- some users may not understand what we mean by "Server", and what the difference is between the status tray "App" running ("Exit KA Lite") and the "Server" running ("Start/Stop Server"). For some users, simply having the app running meaning the server is also running may be simplest, but others will want the control of leaving the status icon running by stopping the server for a while. I feel like there's some magic solution that will make all this clearer, but don't know what it is yet. Any proposals?

**Comment by benjaoming** [comment]
Possible suggestion:

Start KA Lite
Stop KA Lite

Discussion: It could also be argued that understanding the concept of "server" is a prerequisite for running the server. Otherwise people are gonna have a hard time discerning documentation and communicating KA Lite concepts... i.e. "KA Lite doesn't run" as opposed to "My KA Lite server doesn't run".

P.s. looking good, @ruimalheiro, I'll be trying to match the UX for Linux/GTK shortly.

**Comment by kkmach** [comment]
I understand that the concept of a server may be confusing- and thereby being knowledgable of the consequences of turning the server on/off could be arbitrary. I usually turn to the app tray as a shortcut to turn things off/on and I think introducing the option of server control may interrupt this mental model, especially for people who are unfamilar with the concept of a server. I 

I agree with the ben's suggestion. Maybe control the on/off of ka lite with the task tray and then have an icon or indicator in the main gui that shows server state

**Comment by jamalex** [comment]
Yeah, great input. And we also need something to launch the KA Lite URL in the browser, so perhaps something like the below?

`Start KA Lite`
`Stop KA Lite`
`Open KA Lite` (not sure about the wording on this one)
`Options >`
`Exit`

**Comment by jamalex** [comment]
And how should we reword the `Options` submenu items, given this way of thinking about the system?

**Comment by kkmach** [comment]
wouldn't start and open KA Lite mean the same thing? Maybe just Open Ka Lite, Exit KA Lite. Run Options can be Open KA Lite when you log in, Open KA Lite when computer boots, the last option seems fine.

**Comment by jamalex** [comment]
> wouldn't start and open KA Lite mean the same thing?

Nope, they're different. `Start` = "run the server", whereas `Open` = "load the browser with the correct URL". In many cases, only one is desired, for example:

- If KA Lite is installed on this computer to be accessed by other computers over the LAN, there's no need to load the browser.
- If the KA Lite server is already running, but the browser was closed, the user may need a way to load the correct URL in the browser again.

**Comment by benjaoming** [comment]
@jamalex "Open KA Lite" could also mean "Start the server if it isn't active, then navigate to the URL in the user's default browser".

However, I think it's also an illustration of the problems that occur when trying to shield the user from being confused by the concept of a server. It could end up being a disservice. It's like what Windows once did... Start -> Shutdown :D

**Comment by jtamiace** [comment]
What about "Load KA Lite" for starting the server, since preparing the server appears to someone who is not familiar with them to be preparing KA Lite to run. 

And after the user clicks that, another button/link could appear/become clickable to allow the user to open KA Lite in the browser. And that can be something like "Open in (default browser)" or any of the previously mentioned suggestions.

Also, would it be possible to just have one exit option? If a user stops the server, they may try to continue interacting with the interface and find that it doesn't take them anywhere. If there is just one option to simultaneously stop the server and close the browser tab that KA Lite was running in, it might be a little less confusing in terms of determining the difference between "Stop KA Lite" and "Stop server."

**Comment by jtamiace** [comment]
So my suggested gui options would be:

Before starting server:
Load KA Lite
Open in browser <--- not yet clickable
Options
Exit KA Lite

After starting server:
Load KA Lite <---- no longer clickable (??)
Open in browser <---- now clickable 
Options
Exit KA Lite

**Comment by jamalex** [comment]
Thanks for chiming in, @jtamiace!

That would be another way to think about it (and how some programs behave): only having the task tray icon appear when KA Lite (server and all) is actually running. The disadvantage is that once they close KA Lite, they wouldn't have convenient access to the task tray icon for restarting it. Also, it's probably not feasible to try to kill KA Lite browser windows upon exiting, so that problem would still exist.

The push-and-pull in the design here probably comes from there being two different competing use cases:

- KA Lite being run as a *server* for other computers on the LAN
- KA Lite being run as an *app*, where somebody wants to  just use it, locally.

Since we don't know how it will be used, in advance, we probably just need to find a decent compromise that makes sense for both.

**Comment by crystalkwok** [comment]
Hi guys!  Sorry if I'm not understanding it clearly, since I'm still unsure
about the technical side of things, but from what I'm seeing, I think
Jessica's approach seems pretty good.  for the server, I'd either go with
"Load KA Lite" or "Start Up KA Lite", and I would also go with Jessica's
approach to "Open KA Lite" for opening in browser. To me, the language to
use "Close KA Lite" (to close the app) and "Exit (or Quit) KA Lite" (to
close the server) makes sense.

This might be thinking too far, considering ive never actually seen it done
before, but could there be like a light notification next to the app tray
icon or in the icon (green for the server is up and running, red for its
not), and it will change accordingly?  Then the user would know if the
server is up and running and if they can click on the app tray icon or not.


Again, I'm sorry if I'm not completely understanding it!

**Comment by benjaoming** [comment]
I'd recommend against having UI elements dynamically manipulated since the true state of KA Lite is known in the server itself, less in the CLI, and even less in the UI programme. Thus, there is a substantial possibility of inconsistency in this regard. Well, this would at least require a new command `kalite status` or something. I'm putting that in the learningequality/kalite#2757

**Comment by jamalex** [comment]
Yes, there was some discussion of status-checking from the GUI on the Windows repo:
https://github.com/learningequality/ka-lite-installer-windows/issues/37#issuecomment-70409242

Showing this status in the GUI seems pretty important. Perhaps some part of the icon should change from red (stopped) to orange (starting) to green (running)?

**Comment by kkmach** [comment]
maybe should just keep it red from stopped and green for running. Orange
kind of reminds me of an idle state. Also need to factor in how long KA
Lite server to boot up and run. Not so sure about the technical aspects of
this, but if it is a long wait, having the orange starting indicator would
be confusing

**Comment by ruimalheiro** [comment]
Thanks a lot @benjaoming ! Looking forward to see the Linux/GTK version!

Why not just having: 

- Load in browser;
- options -> auto start when user logs in | auto start when windows boots up;
- Exit KA Lite

I think we can assume that once KA Lite GUI is run, the server starts everytime. Why do we need to actually control the server? Updates? An installer can find if KA Lite is running and start and stop it as needed. Also the GUI can catch failures and reboot the server as needed.

That way, any user would assume that once the GUI is running, the server is up.

(More advanced but we could have a thread checking for updates and asking if the user wants to update if there is anything available :) )

Anyway, just ideas :+1:

**Comment by benjaoming** [comment]
Hi @ruimalheiro :)

Thanks!! Looking forward to having some more talks about our packaging adventures :)

> Also the GUI can catch failures and reboot the server as needed.

I think that's putting too much functionality in the UI part because it's platform independent as things are looking atm... but you could put it somewhere else. However, it poses a new challenge, namely that a server can start and stop forever due to a bug... and then eventually the phone doesn't have anymore battery :) I think that if the server actually dies, then it should be allowed to, and then the UI should somehow be able to express that state... either with log output or by querying once in a while about the status of the server.

I wouldn't mind being able to start and stop the server explicitly. As a dev or someone supporting a user, I would like to say "press stop, then press start, then check the log output" or something like that.

**Comment by benjaoming** [comment]
Agree with @kkmach that an "orange state" will be difficult, especially because it might occur in a situation where the server cannot be queried for ambiguous reasons... so it can either mean that it hasn't started or just that it's busy :)

**Comment by ruimalheiro** [comment]
Hey @benjaoming !! (Sorry for the late reply :+1: ) Anytime! :dancer: :D

Yeah, I think I agree with that. The problem here is that not everyone have the notion of control (start/stop) and many people are used to think that when something is in the system tray, then, it is running already. 

Technically, is way better for us to use and offer that control indeed :+1:

**Comment by cpauya** [comment]
Do we now have icons I can use for the [Mac OSX .app file](https://github.com/learningequality/installers/issues/3) system menu?

Refer to @benjaoming's [comment regarding useful statuses](https://github.com/learningequality/installers/issues/3#issuecomment-72085613) when the .app is run.

Specifically, we need icons when app status is:
* launched but not started - perhaps a logo icon?
* running
* stopped
* error or needs user attention

/cc: @jtamiace @66eli77

**Comment by jamalex** [comment]
Not sure we really arrived at a finalized version on this. Can someone post a proposal based on everything here, so all the platform developers can standardize on that?

Another thing that would be useful is showing the current system IP address(es); this has been a recurring theme in many deployments, so highlighting it here would be great.

**Comment by benjaoming** [comment]
The Ubuntu/Debian part of this issue is tracked here (although it hasn't been worked on)

https://github.com/learningequality/ka-lite-gtk/issues/3

Aside from that, I guess that since both OSX and Windows have task tray / menu items, we should close this issue and open up separate issues regarding specific bugs or design changes.
